### PR TITLE
feat: add open PR comments post

### PR DESCRIPTION
**Pull request by cathteng**

### CODE REVIEW

File: data/blog/how-open-pr-comments-work.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
+---
+title: 'How open PR comments work'
+date: '2024-04-18'
+tags: ['integrations', 'github', 'pr-comments']
+draft: false
+summary: "This is how we manage to comment Sentry issues related to code you're modifying in a pull request within seconds."
+images: []
```
Comment (cathteng) [inline data/blog/how-open-pr-comments-work.md:7]: missing a fun image

File: data/blog/how-open-pr-comments-work.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+Why do we do this? At a certain point, when a PR is touching so many lines and/or files, it becomes less and less useful to point out specific issues related to the functions inside the PR. For instance, if somebody applies a linting change, it could result in a lot of lines modified.
+
+Note that we only support a particular set of languages (via looking at file extensions). We skip counting files and lines if they don’t have a file extension we support, and only continue with the files that we can support if these checks are met.
+
+### Fetching issues for each file
+
+#### Reverse codemapping
+
+1.  Normal codemappings map a file in the stack trace (within Sentry) to the source code (in a source code management integration such as Github) using a stack trace root and a source code root. Sentry may store filenames differently than in Github, so codemappings store this relationship. We leverage codemappings here because we want to figure out the Sentry project(s) associated with the file and the stored name of the file in order to make a Snuba query, which requires `project_id`.
```
Comment (iamrajjoshi) [inline data/blog/how-open-pr-comments-work.md:43]: Do we need to explain Snuba or add a link to what it is?

**Review by iamrajjoshi — APPROVED** [review]
🚢

File: data/blog/how-open-pr-comments-work.md (hunk 2 above)
Comment (JoshFerge) [reply data/blog/how-open-pr-comments-work.md:43]: definitely!

Comment (bruno-garcia) [inline data/blog/how-open-pr-comments-work.md:43]: _codemappings map_ -> I work here and I'm not sure what this is, how about linking to the docs?

Comment (bruno-garcia) [reply data/blog/how-open-pr-comments-work.md:43]: added a suggestion below

File: data/blog/how-open-pr-comments-work.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+2.  We use the organization, the repository, and the filename to attempt to fetch code mappings for the file, matching on whether any source code root for a codemapping is a substring of the filename. If any codemappings are found, we reverse codemap by replacing the source code root in the filename from Github with the stack trace root. (Reverse because usually we go from stack trace to source code root, for instance when opening up a line from a stack trace in Github)
+
+#### Extract functions from the file patch
+
+1.  Depending on the language of the file, we fetch the appropriate parser and apply it to the file patch (the git diff for the file).
```
Comment (bruno-garcia) [inline data/blog/how-open-pr-comments-work.md:48]:
```suggestion
1.  Depending on the language of the file, we fetch the appropriate parser and apply it to the file patch (the `git diff` for the file).
```

File: data/blog/how-open-pr-comments-work.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+### Create comment
+
+Then for each file, we make a little table with the function and issue information. All the tables besides the one for the first file are hidden in a toggle. Each file type may also have a slightly different formatting template. In Javascript/Typescript, we want to show the `Affected Users` because it’s more important for frontend. Meanwhile, Python usually always has 0 affected users so it’s not shown.
+
+![Javascript open PR comment](/images/open-pr-comments/open-pr-comment.png)
+
+## A note on language parsers
+
+We support different languages in open PR comments. However, they also require different methods to extract functions from the diff, different handling for the ways events with those functions might be stored in Snuba, and how to format the comment table for each file type.
```
Comment (bruno-garcia) [inline data/blog/how-open-pr-comments-work.md:68]: Oh I guess the answer is here. Should we mention which ones or link to the docs so it evolves with time?

**Review by bruno-garcia — COMMENTED** [review]
Mostly fixing casing but some other suggestions.

Also note it's fair to llink to actual code. If you link to a commit sha, selecting a line range will be stable, for example when you talk about the regex on the file and git hunk headers. Not a must but I find that code explains better than words sometimes.

File: data/blog/how-open-pr-comments-work.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+- Issue row template for the comment
+- Extracting functions from patch
+- [multiIf](https://clickhouse.com/docs/en/sql-reference/functions/conditional-functions#multiif)
+  - This can contain custom logic to fetch the function name from the stackframe that matches the filename + a name within the list of function names.
+  - We do this because we can match up to X frames deep in the stacktrace, we might have a set of function names we’re matching on, and we want the actual function name that we matched on in the stack trace. The stack trace is stored as an array.
+
+### Example language parser for Javascript
+
+```
+class JavascriptParser(LanguageParser):
```
[the remark this answers is not in the document]
Comment (cathteng) [reply data/blog/how-open-pr-comments-work.md:83]: i can link to the actual code

File: data/blog/how-open-pr-comments-work.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+#### Extract functions from the file patch
+
+1.  Depending on the language of the file, we fetch the appropriate parser and apply it to the file patch (the git diff for the file).
+2.  This applies a regex to the whole file, looking for git hunk headers (e.g. `@@ -188,9 +188,7 @@ def __init__():` for Python) that indicate a function in the language, and that the code modified in the section below it belongs to that function. This is more or less correct in finding functions being modified in a PR, unless the function is super short or the lines modified are near the top of the function (this is a limitation of git).
+3.  For Python, the regex only looks for `def {function_name}.` For Javascript/Typescript, there are more ways to initialize a function (function declaration, arrow function, function expression, etc), so there are multiple regexes to find function names.
```
[the remark this answers is not in the document]
Comment (cathteng) [reply data/blog/how-open-pr-comments-work.md:50]: we also support ruby and PHP, i can link to docs. we don't have an active GH issue to track requests

File: data/blog/how-open-pr-comments-work.md (hunk 4 above)
Comment (cathteng) [reply data/blog/how-open-pr-comments-work.md:68]: yeah i'll link to docs

File: data/blog/how-open-pr-comments-work.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+## Comment workflow
+
+### Qualification checks
+
+We fetch information for the PR files from Github, and then ensure that we don’t comment on PRs that have any of the following:
+
+1. More than 7 files modified
+   - We cannot extract function information from deleted files because the Github API does not give us a patch, and it doesn’t make sense to count new files since they won’t have any issues associated with them
+2. More than 500 lines changed
```
[the remark this answers is not in the document]
Comment (cathteng) [reply data/blog/how-open-pr-comments-work.md]: i broke it up so it reads easier

File: data/blog/how-open-pr-comments-work.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (bruno-garcia) [reply data/blog/how-open-pr-comments-work.md]: We can create a ticket and link here too, that works

[the remark this answers is not in the document]
Comment (cathteng) [reply data/blog/how-open-pr-comments-work.md]: i link the GH discussion at the bottom, is that fine? https://github.com/getsentry/sentry.engineering/pull/125/files#diff-af626cdddc04b347356f49d301f74056c1ce1757614f238f254a8b293fed81d4R102

**Comment by cathteng** [comment]
@bruno-garcia how do i get cool artwork for my blog post? should i just find an image that already exists?

**Comment by bruno-garcia** [comment]
> @bruno-garcia how do i get cool artwork for my blog post? should i just find an image that already exists?

You can use an existing image if you'd like to expedite this. Or put a request to the design team if you think this needs a custom image.

Maybe there's an image from an old post on this blog you can just reuse?

File: data/blog/how-open-pr-comments-work.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (bruno-garcia) [reply data/blog/how-open-pr-comments-work.md]: Sure, before reading this I ended up creating a ticket specificaly for languages though, added to the docs too: https://github.com/getsentry/sentry-docs/pull/9837

**Review by bruno-garcia — APPROVED** [review]
LGTM!

**Comment by bruno-garcia** [comment]
@cathteng :shipit: 
Great post! No strong feelings on the image specially if it's an existing Sentry specific one

**Pull request merged**